### PR TITLE
[Clang] prevent setting default lexical access specifier for missing primary declarations

### LINKED ISSUE #112208: [clang++] Assertion `Access != AS_none && "Access specifier is AS_none inside a record decl"' failed in clang::Decl::AccessDeclContextCheck() const
**This testcase is generated by a fuzzer.**

Compiler Explorer: [https://godbolt.org/z/rv1h188aq](https://godbolt.org/z/rv1h188aq)

This invalid code will crash on Clang Assertion Trunk, it seems that this goes back to clang-3.1:
```cpp
class C {
    enum E { e1=0 };
    void fun1(int, enum E;);
};
```

Stack dump:
```
clang++: /root/llvm-project/clang/lib/AST/DeclBase.cpp:1106: bool clang::Decl::AccessDeclContextCheck() const: Assertion `Access != AS_none && "Access specifier is AS_none inside a record decl"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/clang++ -gdwarf-4 -g -o /app/output.s -mllvm --x86-asm-syntax=intel -fno-verbose-asm -S --gcc-toolchain=/opt/compiler-explorer/gcc-snapshot -fcolor-diagnostics -fno-crash-diagnostics <source>
1.	<source>:3:26: current parser token ';'
2.	<source>:1:1: parsing struct/union/class body 'C'
 #0 0x0000000003ba9d48 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3ba9d48)
 #1 0x0000000003ba7a0c llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3ba7a0c)
 #2 0x0000000003af4c68 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x000071deb5042520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x000071deb50969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x000071deb5042476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x000071deb50287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x000071deb502871b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)
 #8 0x000071deb5039e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
 #9 0x0000000007a353c0 clang::Decl::AccessDeclContextCheck() const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x7a353c0)
#10 0x000000000681e85a clang::Sema::SetMemberAccessSpecifier(clang::NamedDecl*, clang::NamedDecl*, clang::AccessSpecifier) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x681e85a)
#11 0x00000000069e7375 clang::Sema::ActOnTag(clang::Scope*, unsigned int, clang::TagUseKind, clang::SourceLocation, clang::CXXScopeSpec&, clang::IdentifierInfo*, clang::SourceLocation, clang::ParsedAttributesView const&, clang::AccessSpecifier, clang::SourceLocation, llvm::MutableArrayRef<clang::TemplateParameterList*>, bool&, bool&, clang::SourceLocation, bool, clang::ActionResult<clang::OpaquePtr<clang::QualType>, false>, bool, bool, clang::Sema::OffsetOfKind, clang::SkipBodyInfo*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x69e7375)
#12 0x000000000668acf8 clang::Parser::ParseEnumSpecifier(clang::SourceLocation, clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, clang::Parser::DeclSpecContext) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x668acf8)
#13 0x0000000006687a76 clang::Parser::ParseDeclarationSpecifiers(clang::DeclSpec&, clang::Parser::ParsedTemplateInfo&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::Parser::LateParsedAttrList*, clang::ImplicitTypenameContext) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6687a76)
#14 0x000000000668e0d7 clang::Parser::ParseParameterDeclarationClause(clang::DeclaratorContext, clang::ParsedAttributes&, llvm::SmallVectorImpl<clang::DeclaratorChunk::ParamInfo>&, clang::SourceLocation&, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x668e0d7)
#15 0x000000000668f438 clang::Parser::ParseFunctionDeclarator(clang::Declarator&, clang::ParsedAttributes&, clang::BalancedDelimiterTracker&, bool, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x668f438)
#16 0x000000000669454f clang::Parser::ParseDirectDeclarator(clang::Declarator&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x669454f)
#17 0x0000000006681586 clang::Parser::ParseDeclaratorInternal(clang::Declarator&, void (clang::Parser::*)(clang::Declarator&)) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6681586)
#18 0x00000000067b6251 clang::Sema::runWithSufficientStackSpace(clang::SourceLocation, llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x67b6251)
#19 0x0000000006670baf clang::Parser::ParseDeclarator(clang::Declarator&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6670baf)
#20 0x00000000066a5745 clang::Parser::ParseCXXMemberDeclaratorBeforeInitializer(clang::Declarator&, clang::VirtSpecifiers&, clang::ActionResult<clang::Expr*, true>&, clang::Parser::LateParsedAttrList&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x66a5745)
#21 0x00000000066ac7c0 clang::Parser::ParseCXXClassMemberDeclaration(clang::AccessSpecifier, clang::ParsedAttributes&, clang::Parser::ParsedTemplateInfo&, clang::ParsingDeclRAIIObject*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x66ac7c0)
#22 0x00000000066ae4b1 clang::Parser::ParseCXXClassMemberDeclarationWithPragmas(clang::AccessSpecifier&, clang::ParsedAttributes&, clang::TypeSpecifierType, clang::Decl*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x66ae4b1)
#23 0x00000000066afff7 clang::Parser::ParseCXXMemberSpecification(clang::SourceLocation, clang::SourceLocation, clang::ParsedAttributes&, unsigned int, clang::Decl*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x66afff7)
#24 0x00000000066b27a3 clang::Parser::ParseClassSpecifier(clang::tok::TokenKind, clang::SourceLocation, clang::DeclSpec&, clang::Parser::ParsedTemplateInfo&, clang::AccessSpecifier, bool, clang::Parser::DeclSpecContext, clang::ParsedAttributes&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x66b27a3)
#25 0x0000000006686fa0 clang::Parser::ParseDeclarationSpecifiers(clang::DeclSpec&, clang::Parser::ParsedTemplateInfo&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::Parser::LateParsedAttrList*, clang::ImplicitTypenameContext) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6686fa0)
#26 0x0000000006644494 clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6644494)
#27 0x0000000006644efe clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6644efe)
#28 0x000000000664c693 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x664c693)
#29 0x000000000664d57d clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x664d57d)
#30 0x000000000664da20 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x664da20)
#31 0x000000000663fa62 clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x663fa62)
#32 0x0000000004502248 clang::CodeGenAction::ExecuteAction() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x4502248)
#33 0x00000000047bbb49 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x47bbb49)
#34 0x000000000473a6ee clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x473a6ee)
#35 0x00000000048a16be clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x48a16be)
#36 0x0000000000cdb8ef cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xcdb8ef)
#37 0x0000000000cd35fa ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#38 0x0000000004545f19 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#39 0x0000000003af5114 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3af5114)
#40 0x000000000454650f clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#41 0x000000000450be4d clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x450be4d)
#42 0x000000000450cf3d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x450cf3d)
#43 0x0000000004514865 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x4514865)
#44 0x0000000000cd87bf clang_main(int, char**, llvm::ToolContext const&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xcd87bf)
#45 0x0000000000ba7d14 main (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xba7d14)
#46 0x000071deb5029d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#47 0x000071deb5029e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#48 0x0000000000cd30ae _start (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xcd30ae)
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
Compiler returned: 134
```

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #112208 

---

This PR resolves a crash triggered by a forward reference to an enum type in a function parameter list. The fix includes setting `Invalid` when `TagUseKind` is `Declaration` to ensure correct error handling.

### CODE REVIEW

File: clang/lib/Sema/SemaAccess.cpp [hunk 1]
```diff
@@ -39,7 +39,8 @@ bool Sema::SetMemberAccessSpecifier(NamedDecl *MemberDecl,
                                     AccessSpecifier LexicalAS) {
   if (!PrevMemberDecl) {
     // Use the lexical access specifier.
-    MemberDecl->setAccess(LexicalAS);
```
Comment (a-tarasyuk) [inline clang/lib/Sema/SemaAccess.cpp:42]: @shafik https://github.com/llvm/llvm-project/issues/112208#issuecomment-2411881773 when `PrevMemberDecl` exists but isn't equal to `LexicalAS` — with `AS_none` excluded

https://github.com/llvm/llvm-project/blob/020566701030425f44eb80387d0ae76c5a867aa9/clang/lib/Sema/SemaAccess.cpp#L48

 I suppose `AS_none` should be excluded when `PrevMemberDecl` is unknown. What do you think?

Comment (shafik) [reply clang/lib/Sema/SemaAccess.cpp:42]: I feel like we should be catching this in `Sema::ActOnTag` here:

https://github.com/llvm/llvm-project/blob/b49701085414838deb0213b9f10b68d9d3af6b0a/clang/lib/Sema/SemaDecl.cpp#L17938-L17951

It looks like we are not b/c `TUK` is `TUK_Declataion` but we should not be forward declaring here either. 

The current diagnostic does not fit this case though and maybe it has to be tweaked some more.

CC @AaronBallman

Comment (shafik) [reply clang/lib/Sema/SemaAccess.cpp:42]: Please also add more details to the summary, it is helpful to have some idea what you think the underlying problem is and why this fix is correct before reviewing.

**Comment by a-tarasyuk** [comment]
@shafik could you take a look at the latest changes? thanks

File: clang/lib/Sema/SemaAccess.cpp (hunk 1 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaAccess.cpp]: @shafik could you take a look at the latest changes? thanks

File: clang/lib/Sema/SemaDecl.cpp [hunk 2]
```diff
@@ -17957,6 +17957,8 @@ Sema::ActOnTag(Scope *S, unsigned TagSpec, TagUseKind TUK, SourceLocation KWLoc,
             << Name;
         Invalid = true;
       }
+      if (TUK == TagUseKind::Declaration && Name)
```
Comment (shafik) [inline clang/lib/Sema/SemaDecl.cpp:17960]: Why do we need to check `Name` as well?

File: clang/lib/Sema/SemaAccess.cpp (hunk 1 above)
Comment (shafik) [reply clang/lib/Sema/SemaAccess.cpp]: I think these changes make sense, the diagnostic on forward reference is not great but that feels like a seperate already existing problem and not directly related to the crash.

Please update the summary. I also had one question on why we are checking `Name` as well.

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaAccess.cpp]: @shafik thanks for the feedback. I've checked, and it seems safe to omit `Name` here, so I’ve removed it.

**Review by shafik — APPROVED** [review]
Assuming it passes the test LGTM

**Comment by a-tarasyuk** [comment]
@shafik thanks for the review. It seems the `Test documentation build` failure isn't related to these changes.

**Comment by a-tarasyuk** [comment]
@shafik if no further review is needed, could you please merge it? I don’t have permission to do it myself. thanks

File: clang/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -585,10 +585,11 @@ Bug Fixes to C++ Support
 - Clang incorrectly considered a class with an anonymous union member to not be
   const-default-constructible even if a union member has a default member initializer.
   (#GH95854).
-- Fixed an assertion failure when evaluating an invalid expression in an array initializer. (#GH112140)
+- Fixed an assertion failure when evaluating an invalid expression in an array initializer (#GH112140)
```
Comment (zyn0217) [inline clang/docs/ReleaseNotes.rst:588]: nit: Why drop the dot here?

Comment (a-tarasyuk) [reply clang/docs/ReleaseNotes.rst]: @zyn0217 oops, it looks like it was removed during the merge conflict resolution. I've reverted it — thanks.

**Comment by zyn0217** [comment]
Thanks!

**Pull request merged**